File: RestroomFinder/Native/Apple/Restroom/AppStorage.swift
//
//  AppStorage.swift
//  Restroom
//
//  Created by DevTeam on 2/22/16.
//  Copyright © 2016 DevTeam. All rights reserved.
//

import Foundation
import SharedFramework
class AppStorage{
    static let disclaimerKey = "Disclaimer_Agreed"
    static let disclaimerPopupKey = "DisclaimerPopup_Agreed"
    
    static var _keyValues: [KeyValue]!
    
    static func shouldPopupDisclaimer() -> Bool{
        let keyvalue = keyValues?.tryGetValue(disclaimerPopupKey)
        
        if (keyvalue == nil){
            return true
        }
        else {
            //print ("now: \(DateTime.now().toString()) vs \(keyvalue?.updateDate?.toString())")
            let now = DateTime.now()
            return (keyvalue?.updateDate?.year != now.year) ||
                    (keyvalue?.updateDate?.month != now.month) ||
                    (keyvalue?.updateDate?.day != now.day)
        }
    }

    static func setPopupDisclaimer(){
        var res = keyValues
        if (res == nil){
            res = [KeyValue]()
        }
        var keyvalue = res?.tryGetValue(disclaimerPopupKey)
        if (keyvalue == nil){
            keyvalue = KeyValue(key: disclaimerPopupKey, value: "set", updateDate: DateTime.now())
            res?.append(keyvalue!)
        }
        else {
            keyvalue?.updateDate = DateTime.now()
        }
        invalidateCache()
        KeyValue.Save(res!)
    }
    
    
    static func isFirstTimeRunningApplication() -> Bool{
        //return true
        if (keyValues == nil){
            return true
        }
        return false
    }
    
    static func setFirstTimeRunningApplication(badge: String) {
        var res = keyValues
        if (res == nil){
            res = [KeyValue]()
        }
        let keyvalue = res?.tryGetValue(disclaimerKey)
        if (keyvalue == nil){
            res?.append(KeyValue(key: disclaimerKey, value: badge, updateDate: DateTime.now())!)
        }
        else{
            keyvalue?.updateDate = DateTime.now()
            keyvalue?.value = badge
        }
        invalidateCache()
        KeyValue.Save(res!)
    }
    
    static var badge : String? {
        get{
            return keyValues?.tryGetValue(disclaimerKey)?.value
        }
    }


    
    private static var keyValues: [KeyValue]? {
        get{
            if (_keyValues == nil){
                _keyValues = KeyValue.Load()
            }
            return _keyValues
        }
    }
    private static func invalidateCache(){
        _keyValues = nil;
    }
}

File: RestroomFinder/Native/Android/RestroomFinder/app/src/main/java/org/actransit/restroomfinder/Infrastructure/AppStorage.java
package org.actransit.restroomfinder.Infrastructure;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;

import org.actransit.restroomfinder.Model.KeyValue;
import org.json.JSONObject;

import java.util.Calendar;


/**
 * Created by DevTeam on 6/21/16.
 */
public class AppStorage extends Activity{
    private static final String PREFS_NAME = "RestroolFinderFile";
    private static final String disclaimerKey = "Disclaimer_Agreed";
    private static final String disclaimerPopupKey = "DisclaimerPopup_Agreed";
    private static SharedPreferences settings;
    private static AppStorage current;
    private AppStorage(Context context){
        settings= context.getSharedPreferences(PREFS_NAME, 0);
    }

    private KeyValue badge;


    public static AppStorage Current(Context context){
        if (current!=null)
            return current;
        current=new AppStorage(context);
        return current;
    }

    public boolean shouldPopupDisclaimer() {
        String value=settings.getString(disclaimerPopupKey,"");
        if (value=="")
            return true;
        try{
            JSONObject jObj=new JSONObject(value);
            KeyValue keyValue=JSONMapper.toObject(jObj, KeyValue.class);
            Calendar now=Calendar.getInstance();
            Calendar c= Calendar.getInstance();
            c.setTime(keyValue.updateDate);
            return c.get(Calendar.YEAR) != now.get(Calendar.YEAR) ||
                    c.get(Calendar.MONTH) != now.get(Calendar.MONTH) ||
                    c.get(Calendar.DAY_OF_MONTH) != now.get(Calendar.DAY_OF_MONTH);

        }
        catch (Exception e){
            e.printStackTrace();
            return true;
        }
    }

    public void setPopupDisclaimer(){
        saveItem(disclaimerPopupKey,"set");
    }


    public boolean isFirstTimeRunningApplication(){
        String disClaimerAgreed = settings.getString(disclaimerKey, "");
        return disClaimerAgreed=="";// | true;
    }

    public void setFirstTimeRunningApplication(String badge) {
        saveItem(disclaimerKey,badge);
    }

    public KeyValue getBadge(){
        if (badge==null)
            badge=restoreItem(disclaimerKey);
        return badge;
    }


    public void saveItem(String key,String value){
        try{
            invalidateCache();
            JSONObject jObj=new JSONObject();
            jObj.put("key",key);
            jObj.put("value",value);
            jObj.put("updateDate",Calendar.getInstance().getTime().getTime());
            String v = jObj.toString();
            SharedPreferences.Editor editor = settings.edit();
            editor.putString(key, v);
            editor.commit();

        }
        catch(Exception e){
            e.printStackTrace();
        }
    }
    public KeyValue restoreItem(String key) {
        String value=settings.getString(key,"");
        if (value=="")
            return null;
        try{
            JSONObject jObj=new JSONObject(value);
            KeyValue keyValue=JSONMapper.toObject(jObj, KeyValue.class);
            return keyValue;
        }
        catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }

    private void invalidateCache(){
        badge=null;
    }

}
